### PR TITLE
add basic attributes test

### PR DESCRIPTION
**Pull request by Zeskbest**

### CODE REVIEW

File: tests/test_estimation_basic_attributes.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    Generate warps.
+
+    Args:
+        faceEngine: Face Engine instance
+        imagePath: path to image
+
+    Returns:
+        first 1000 warps sorted by coordinates
+    """
+    detector = faceEngine.createFaceDetector(DetectorType.FACE_DET_V3)
```
Comment (matemax) [inline tests/test_estimation_basic_attributes.py]: use default

File: tests/test_estimation_basic_attributes.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+                    try:
+                        if isinstance(singleAttr, Ethnicities):
+                            self.assertEqual(singleAttr.asian, batchAttr.asian, msg)
+                            self.assertEqual(singleAttr.indian, batchAttr.indian, msg)
+                            self.assertEqual(singleAttr.caucasian, batchAttr.caucasian, msg)
+                            self.assertEqual(singleAttr.africanAmerican, batchAttr.africanAmerican, msg)
+                        else:
+                            self.assertEqual(singleAttr, batchAttr, msg)
+                    except AssertionError:
+                        warp.warpedImage.save(filename)
```
Comment (matemax) [inline tests/test_estimation_basic_attributes.py]: for what?

File: tests/test_estimation_basic_attributes.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+                        raise
+
+                    self.assertAlmostEqual(expectedAttr, singleAttr, delta=delta)
```
Comment (matemax) [inline tests/test_estimation_basic_attributes.py]: Need following tests:

1. test on as_dict
2. aggregate parameters
3. batch estimation.
4.  test on parameters estimateAge, estimateGender, estimateEthnicity

File: tests/test_estimation_basic_attributes.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+                    self.assertNotIn(None, (singleAttr, batchAttr))
+
+                    filename = f"./{time()}.jpg"
+                    msg = f"Batch estimation '{estimationType}' differs from single one. Saved as '{filename}'."
+                    try:
+                        if isinstance(singleAttr, Ethnicities):
+                            self.assertEqual(singleAttr.asian, batchAttr.asian, msg)
+                            self.assertEqual(singleAttr.indian, batchAttr.indian, msg)
+                            self.assertEqual(singleAttr.caucasian, batchAttr.caucasian, msg)
+                            self.assertEqual(singleAttr.africanAmerican, batchAttr.africanAmerican, msg)
```
Comment (matemax) [inline tests/test_estimation_basic_attributes.py]: use pytest assert

**Review by matemax — APPROVED** [review]

File: tests/test_estimation_basic_attributes.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        for estimationType, delta in (("Age", AGE_DELTA), ("Gender", GENDER_DELTA), ("Ethnicity", ETH_DELTA)):
+            estimationFlag = f"estimate{estimationType}"
+            basicAttributeGetter: Callable[[BasicAttributes], Union[Ethnicities, float, None]] = attrgetter(
+                estimationType.lower()
+            )
+            for warpNumber, warp in enumerate(self._warps):
+                expectedAttr = getattr(warpNumberToEstimations[warpNumber], estimationType)
+                with self.subTest(estimationType=estimationType, warpNumber=warpNumber):
+                    singleAttr = basicAttributeGetter(self.estimate(warp, **{estimationFlag: True}))
+                    batchAttr = basicAttributeGetter(self.estimateBatch([warp], **{estimationFlag: True})[0][0])
```
Comment (matemax) [inline tests/test_estimation_basic_attributes.py]: validate output type

**Review by matemax — APPROVED** [review]

File: tests/test_estimation_basic_attributes.py (hunk 3 above)
Comment (Zeskbest) [reply tests/test_estimation_basic_attributes.py]: 3 ad 4 done already
added tests:
test_aggregation
test_as_dict

File: tests/test_estimation_basic_attributes.py (hunk 4 above)
Comment (Zeskbest) [reply tests/test_estimation_basic_attributes.py]: ok

File: tests/test_estimation_basic_attributes.py (hunk 5 above)
Comment (Zeskbest) [reply tests/test_estimation_basic_attributes.py]: ok

File: tests/test_estimation_basic_attributes.py (hunk 2 above)
Comment (Zeskbest) [reply tests/test_estimation_basic_attributes.py]: removed

File: tests/test_estimation_basic_attributes.py (hunk 1 above)
Comment (Zeskbest) [reply tests/test_estimation_basic_attributes.py]: applied

File: tests/test_estimation_basic_attributes.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+Estimation = namedtuple("Estimation", ("Age", "Gender", "Ethnicity"))
+
+
+class TestBasicAttributes(BaseTestClass):
+    """ Test basic attributes. """
+
+    estimator: BasicAttributesEstimator = BaseTestClass.faceEngine.createBasicAttributesEstimator()
+
+    _warp: FaceWarpedImage
```
Comment (gluhar2006) [inline tests/test_estimation_basic_attributes.py:59]: comments?

File: tests/test_estimation_basic_attributes.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            )
+            with self.subTest(estimationType):
+                raw = list(map(estimationGetter, estimations))
+                aggregated = estimationGetter(aggregatedEstimation)
+                if isinstance(aggregated, Ethnicities):
+                    self.assertEth(Eth.fromSeveral(raw), aggregated)
+                else:  # age or gender
+                    assert int(mean(raw)) == int(aggregated)
+
+    def test_as_dict(self):
```
Comment (gluhar2006) [inline tests/test_estimation_basic_attributes.py:180]: ethnicites_as_dict?

File: tests/test_estimation_basic_attributes.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        """
+        Test asDict method.
+        """
+        raw, aggregated = self.estimateBatch(
+            [self._warp, self._warp2], estimateAge=True, estimateGender=True, estimateEthnicity=True, aggregate=True
+        )
+        for estimation in [*raw, aggregated]:
+            assert isinstance(estimation.asDict(), dict)
```
Comment (gluhar2006) [inline tests/test_estimation_basic_attributes.py]: any negative test(s)?

Comment (Zeskbest) [reply tests/test_estimation_basic_attributes.py]: nope =) fsdk has only features, not bugs)

File: tests/test_estimation_basic_attributes.py (hunk 7 above)
Comment (Zeskbest) [reply tests/test_estimation_basic_attributes.py:180]: nope, addedjsonschema

File: tests/test_estimation_basic_attributes.py (hunk 6 above)
Comment (Zeskbest) [reply tests/test_estimation_basic_attributes.py:59]: ok

**Pull request merged**